core/src/test/java/org/testcontainers/utility/PrefixingImageNameSubstitutorTest.java

```diff
@@ -0,0 +1,101 @@
+package org.testcontainers.utility;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.rnorth.visibleassertions.VisibleAssertions.assertEquals;
+import static org.testcontainers.utility.PrefixingImageNameSubstitutor.PREFIX_PROPERTY_KEY;
+
+public class PrefixingImageNameSubstitutorTest {
+
+    private TestcontainersConfiguration mockConfiguration;
+    private PrefixingImageNameSubstitutor underTest;
+
+    @Before
+    public void setUp() {
+        mockConfiguration = mock(TestcontainersConfiguration.class);
+        underTest = new PrefixingImageNameSubstitutor(mockConfiguration);
+    }
+
+    @Test
+    public void testHappyPath() {
+        when(mockConfiguration.getEnvVarOrProperty(eq(PREFIX_PROPERTY_KEY), any())).thenReturn("someregistry.com/our-mirror/");
+
+        final DockerImageName result = underTest.apply(DockerImageName.parse("some/image:tag"));
```

Comment:
I am curious how Docker handles a case where registry's host is "some".

Or, to make it look more "real": `localhost/image:tag`

Reply:
Interesting - I'll check. I'd assume that there has to be either a dot in the host part or there has to be a port part.

Reply:
OK, just quickly playing with Docker CLI vs `DockerImageName.parse`, and I think the behaviour is the same.

Docker CLI:
* `some/foo` -> Docker Hub, repo: `some/foo`
* `some.com/foo` -> `some.com` registry, repo: `foo`
* `some:1234/foo` -> `some` registry, port 1234, repo: `foo`

`DockerImageName`:
* `some/foo` -> Docker Hub, repo: `some/foo`
* `some.com/foo` -> `some.com` registry, repo: `foo`
* `some:1234/foo` -> `some` registry, port 1234, repo: `foo`

Things get wild with `localhost`, though:

```
$ docker pull localhost/foo
Using default tag: latest
Error response from daemon: Get http://localhost/v2/: dial tcp [::1]:80: connect: connection refused

$ docker pull localhos/foo    # note missing t
Using default tag: latest
Error response from daemon: pull access denied for localhos/foo, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

i.e. I think Docker CLI special-cases `localhost` as a registry name. We don't imitate this in `DockerImageName.parse` but I suspect it's a niche case that we can avoid.

Reply:
This looks like an example of `localhost` being treated specially in registry, and I imagine further searching would find more instances of this logic: https://github.com/moby/moby/blob/bb23f1bf61cb49c5aa3f7df934d79c305afb7c8c/registry/service.go#L153

Reply:
Good finding! Perhaps we should add a similar logic then (and tests)

Reply:
Can we defer or avoid? This is only going to be relevant if people are running a registry on port 80 on localhost, which seems extremely niche.

Reply:
Ok 👍